crates/bevy_ecs/macros/src/lib.rs

```diff
@@ -380,9 +380,14 @@ pub fn derive_system_param(input: TokenStream) -> TokenStream {
         let w = format_ident!("w");
         let s = format_ident!("s");
         if ident != &w && ident != &s {
-            return syn::Error::new_spanned(lt, "invalid lifetime name: expected `'w` or `'s`")
-                .into_compile_error()
-                .into();
+            return syn::Error::new_spanned(
+                lt,
+                r#"invalid lifetime name: expected `'w` or `'s`
+ 'w -- refers to data stored in the World.
+ 's -- refers to data stored in the SystemParam's state.'"#,
+            )
+            .into_compile_error()
+            .into();
```

Comment:
@james7132 do you think this is an improvement?

This usage is called out in the docs, but IMO the best way of teaching this is to just let users do it wrong, then have the error tell them the right way of doing it.

Reply:
Looks much better to me. Thanks!